x/gamm/keeper/migrate.go

```diff
@@ -118,28 +121,51 @@ func (k Keeper) validateRecords(ctx sdk.Context, records []types.BalancerToConce
 		}
 
 		// Ensure the provided balancerPoolId exists and that it is of type balancer
-		pool, err := k.GetPool(ctx, record.BalancerPoolId)
+		balancerPool, err := k.GetPool(ctx, record.BalancerPoolId)
 		if err != nil {
 			return err
 		}
-		poolType := pool.GetType()
+		poolType := balancerPool.GetType()
 		if poolType.String() != "Balancer" {
 			return fmt.Errorf("Balancer pool ID #%d is not of type balancer", record.BalancerPoolId)
 		}
 
 		// Ensure the provided ClPoolId exists and that it is of type concentrated.
 		// If clPoolID is 0, this signals a removal, so we skip this check.
+		var clPool poolmanagertypes.PoolI
 		if record.ClPoolId != 0 {
-			pool, err = k.clKeeper.GetPool(ctx, record.ClPoolId)
+			clPool, err = k.clKeeper.GetPool(ctx, record.ClPoolId)
 			if err != nil {
 				return err
 			}
-			poolType = pool.GetType()
+			poolType = clPool.GetType()
 			if poolType.String() != "Concentrated" {
 				return fmt.Errorf("Concentrated pool ID #%d is not of type concentrated", record.ClPoolId)
 			}
 		}
 
+		// Ensure the balancer pools denoms are the same as the concentrated pool denoms
```

Comment:
Do we have documentation for migration logic? Can we please add it somewhere? Preferably a section in `architecture.md`

Reply:
Was about to come and ask the same thing.

Reply:
I had added a blurb about this in GAMM here

https://github.com/osmosis-labs/osmosis/blob/1844dcc1c66708cad48590e9cae9cbbdb1c25202/x/gamm/README.md#L189-L191

I can go into more detail if this is desired